src/main/java/htsjdk/samtools/util/GZIIndex.java

```diff
@@ -0,0 +1,365 @@
+            }
+            if (Long.signum(uncompressedOffset) == -1) {
+                throw new IllegalArgumentException("negative uncompressed offset: " + uncompressedOffset);
+            }
+            this.compressedOffset = compressedOffset;
+            this.uncompressedOffset = uncompressedOffset;
+        }
+
+        /** Returns the compressed offset (block-offset). */
+        public long getCompressedOffset() {
+            return compressedOffset;
+        }
+
+        /** Returns the uncompressed offset. */
+        public long getUncompressedOffset() {
+            return uncompressedOffset;
+        }
+
+        @Override
+        public String toString() {
+            return String.format("IndexEntry={compressed=%d(0x%x),uncompressed=%d(0x%x)",
+                    compressedOffset, compressedOffset,
+                    uncompressedOffset, uncompressedOffset);
+        }
+
+        @Override
+        public boolean equals(final Object obj) {
+            if (obj == null || !(obj instanceof IndexEntry)) {
+                return false;
+            }
+            final IndexEntry other = (IndexEntry) obj;
+            return compressedOffset == other.compressedOffset
+                    && uncompressedOffset == other.uncompressedOffset;
+        }
+
+        @Override
+        public int hashCode() {
+            return 31 * Long.hashCode(compressedOffset) + Long.hashCode(uncompressedOffset);
+        }
+    }
+
+    // first entry in the mapping (do not write down)
+    private static final IndexEntry FIRST_MAPPING = new IndexEntry(0, 0);
+
+    private final List<IndexEntry> entries;
+
+    // private constructors
+    private GZIIndex(final List<IndexEntry> entries) {
+        this.entries = entries;
+    }
+
+    /**
+     * Gets the number of blocks on the file.
+     *
+     * @return the number of blocks.
+     */
+    public int getNumberOfBlocks() {
+        return entries.size();
+    }
+
+    /**
+     * Gets an unmodifiable list with the index entries.
+     *
+     * @return index entries.
+     */
+    public List<IndexEntry> getIndexEntries() {
+        return Collections.unmodifiableList(entries);
+    }
+
+    /**
+     * Gets the virtual offset for seek with {@link BlockCompressedInputStream#seek(long)}.
+     *
+     * @param uncompressedOffset the file-offset.
+     *
+     * @return virtual offset for {@link BlockCompressedInputStream}.
+     * @see BlockCompressedFilePointerUtil
+     */
+    public long getVirtualOffsetForSeek(final long uncompressedOffset) {
+        if (uncompressedOffset == 0) {
+            return BlockCompressedFilePointerUtil.makeFilePointer(0, 0);
+        }
+
+        // list of entries at the uncompressed-offset (or afterwards)
+        final IndexEntry indexEntry = entries.stream()
+                // filter for only the ones where the uncompressed offset is larger than the requested
+                .filter(entry -> entry.getUncompressedOffset() >= uncompressedOffset)
```

Comment:
I'm confused by this.  It seems like this is finding the first block after the offset we're looking for.  Don't we want to find the last block before the offset we're looking for?   Or am I confused?  

We might want to implement this as a binary search since there could potentially be many entries in a large gzipped file?

Reply:
This needs a test.

Reply:
This was the first implementation without optimization (I was planning to add the binary search once I implemented tests and after your review). I will add tests and implement the binary search in the next commit.

Reply:
And that's correct, the implementation was completely wrong for several reasons. With the binary search and the tests I debug the code and it should work now.